[JBPM-7834] Provide OpenShiftStartupStrategy - WB Integration

rhtevan opened the pull request:
Providing a solution for WorkBench integration with KieServer OpenShiftStartupStrategy by implementing a new ServerTemplateOpenShiftStorage.

Related PRs: 

- https://github.com/kiegroup/kie-wb-common/pull/2319

- https://github.com/kiegroup/droolsjbpm-integration/pull/1611

kiereleaseuser2 commented:
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

kiereleaseuser2 commented:
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

kiereleaseuser2 commented:
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

kiereleaseuser2 commented:
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/main/java/org/kie/server/controller/openshift/storage/ServerTemplateOpenShiftStorage.java:
```diff
@@ -0,0 +1,237 @@
+    static SimpleInMemoryCache storageCache = new SimpleInMemoryCache();
+
+    KieServerStateOpenShiftRepository repo;
+    long cacheTTL = DEFAULT_CACHE_TTL;
```
It is referenced by the Test class at the same package.

https://github.com/rhtevan/droolsjbpm-integration/blob/bcab87096e356c5c2691ebf827ea0dc4d2908182/kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/test/java/org/kie/server/controller/openshift/storage/ServerTemplateOpenShiftStorageTest.java#L137

mswiderski replied:
so protected should be enough I guess...

rhtevan replied:
Agree.

rhtevan replied:
Updated as per requested

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/main/java/org/kie/server/controller/openshift/storage/ServerTemplateOpenShiftStorage.java:
```diff
@@ -0,0 +1,237 @@
+    public ServerTemplate load(String identifier) {
+        ServerTemplate template = null;
+        try {
+            template = ServerTemplateConverter.fromState(repo.load(identifier));
```
Originally, we thought that caching is good for bulk loading methods such as load(), loadKeys(), and it need to tolerate staled data for short period of time. For loading individual KieServerState/ServerTemplate, it would be better not to employing caching.  What do you think?

sutaakar replied:
@rhtevan Wouldn't this approach bring inconsistencies, when bulk load returns different results than load of single template?

rhtevan replied:
Initial investigating shows that the data used by UI rending server template detail is exclusively retrieved by single load method rather than the bulk load method. Therefore, the risk of inconsistency and yielding negative impact is low. At the moment, we have agreed on keeping the caching strategy implement as simple as necessary.  

https://github.com/kiegroup/kie-wb-common/blob/master/kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/ServerManagementBrowserPresenter.java#L167

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-impl/src/test/java/org/kie/server/controller/impl/storage/ServerTemplateStorageTest.java:
```diff
@@ -0,0 +1,351 @@
+package org.kie.server.controller.impl.storage;
```
Fixed

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-impl/src/test/java/org/kie/server/controller/impl/storage/ServerTemplateStorageTest.java:
```diff
@@ -0,0 +1,351 @@
+        assertNotNull(fromStorage);
+
+        assertEquals(serverTemplate.getId(), fromStorage.getId());
+        if (isRequired())
```
Updated as per requested

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-impl/src/test/java/org/kie/server/controller/impl/storage/ServerTemplateStorageTest.java:
```diff
@@ -0,0 +1,351 @@
+        assertEquals(containerSpec.getReleasedId(), spec.getReleasedId());
+        assertEquals(containerSpec.getServerTemplateKey().getId(), spec.getServerTemplateKey().getId());
+
+        if (isRequired())
```
Updated as per requested

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-impl/src/test/java/org/kie/server/controller/impl/storage/ServerTemplateStorageTest.java:
```diff
@@ -0,0 +1,351 @@
+        assertEquals(serverTemplate.getId(), templateKey.getId());
+
+        if (isRequired())
+            assertEquals(serverTemplate.getName(), templateKey.getName());
```
Updated as per requested

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-impl/src/test/java/org/kie/server/controller/impl/storage/ServerTemplateStorageTest.java:
```diff
@@ -0,0 +1,351 @@
+        assertEquals(serverTemplate.getId(), template.getId());
+
+        if (isRequired())
+            assertEquals(serverTemplate.getName(), template.getName());
```
Updated as per requested

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-impl/src/test/java/org/kie/server/controller/impl/storage/ServerTemplateStorageTest.java:
```diff
@@ -0,0 +1,351 @@
+        assertEquals(containerSpec.getReleasedId(), spec.getReleasedId());
+        assertEquals(containerSpec.getServerTemplateKey().getId(), spec.getServerTemplateKey().getId());
+
+        if (isRequired())
```
Updated as per requested

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/main/java/org/kie/server/controller/openshift/storage/ServerTemplateConverter.java:
```diff
@@ -0,0 +1,178 @@
+
+    public static ServerTemplate fromState(@NotNull KieServerState state) {
+        if (state.getConfiguration() == null)
+            throw new IllegalArgumentException("Kie server configuration can not be empty.");
```
Updated as per requested

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/main/java/org/kie/server/controller/openshift/storage/ServerTemplateConverter.java:
```diff
@@ -0,0 +1,178 @@
+        if (kcfg.getConfigItem(KieServerConstants.KIE_DROOLS_SERVER_EXT_DISABLED) != null)
+            capabilities.remove(Capability.RULE.name());
+        if (kcfg.getConfigItem(KieServerConstants.KIE_OPTAPLANNER_SERVER_EXT_DISABLED) != null)
+            capabilities.remove(Capability.PLANNING.name());
```
Updated as per requested

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/main/java/org/kie/server/controller/openshift/storage/ServerTemplateOpenShiftStorage.java:
```diff
@@ -0,0 +1,237 @@
+    private static final Logger logger = LoggerFactory.getLogger(ServerTemplateOpenShiftStorage.class);
+    private static final String SERVER_TEMPLATE_KEYS_CACHE_KEY = "server.template.keys.cache.key";
+    private static final String SERVER_TEMPLATE_CACHE_KEY = "server.template.cache.key";
+    private static final long DEFAULT_CACHE_TTL = 60000L; // 60 seconds
```
Exposed as a system property with default to 60 seconds.

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/test/java/org/kie/server/controller/openshift/storage/ServerTemplateConverterTest.java:
```diff
@@ -0,0 +1,281 @@
+            if (!template.getServerInstance(id).getServerTemplateId().equals(id))
+                fail("Server instance templateId doesn't match up with Id!");
+
+            if (!template.getServerInstance(id).getUrl().equals(url))
```
Updated as per requested

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/test/java/org/kie/server/controller/openshift/storage/ServerTemplateConverterTest.java:
```diff
@@ -0,0 +1,281 @@
+                if (!conSpec.getServerTemplateKey().getId().equals(id))
+                    fail("Server template id within container spec doesn't match!");
+                if (!conSpec.getServerTemplateKey().getName().equals(id))
+                    fail("Server template name within container spec doesn't match!");
```
Updated as per requested

rhtevan commented on kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/test/java/org/kie/server/controller/openshift/storage/ServerTemplateConverterTest.java:
```diff
@@ -0,0 +1,281 @@
+                if (!conResFound.getReleaseId().equals(relId))
+                    fail("Container releaseId doesn't match!");
+                if (!conResFound.getStatus().equals(status))
+                    fail("Container status doesn't match!");
```
Updated as per requested

mswiderski commented on kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/main/java/org/kie/server/controller/openshift/service/OpenShiftServerTemplateStorageService.java:
```diff
@@ -0,0 +1,33 @@
+import org.kie.server.controller.openshift.storage.OpenShiftServerTemplateStorage;
+
+public class OpenShiftServerTemplateStorageService	implements PersistingServerTemplateStorageService {
+	private static OpenShiftServerTemplateStorage storage;
```
formatting seems to be a bit off here

rhtevan replied:
Fixed formatting.

mswiderski commented:
ok to test

sutaakar commented on kie-server-parent/kie-server-controller/kie-server-controller-impl/src/test/java/org/kie/server/controller/impl/storage/ServerTemplateStorageTest.java:
```diff
@@ -0,0 +1,371 @@
+        assertEquals(0, instances.size());
+    }
+
+    protected boolean isRequired() {
```
What does this condition mean? Please adjust the name to be more descriptive.

rhtevan replied:
Renamed the method and added short description. 

https://github.com/rhtevan/droolsjbpm-integration/blob/fd4635e8e9b2263458db3339d7ea0c94b7ae87b6/kie-server-parent/kie-server-controller/kie-server-controller-impl/src/test/java/org/kie/server/controller/impl/storage/ServerTemplateStorageTest.java#L368-L374

sutaakar commented on kie-server-parent/kie-server-controller/kie-server-controller-impl/src/test/java/org/kie/server/controller/impl/storage/ServerTemplateStorageTest.java:
```diff
@@ -0,0 +1,371 @@
+        assertEquals(newContainerSpec.getConfigs().size(), spec.getConfigs().size());
+    }
+
+    protected void createServerTemplateWithContainer() {
```
Is there any reason why this method is called in implementation classes and not here in `@Before` method?

rhtevan replied:
That method is called by concrete test case class at the @Before method. For example,

https://github.com/rhtevan/droolsjbpm-integration/blob/fd4635e8e9b2263458db3339d7ea0c94b7ae87b6/kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/test/java/org/kie/server/controller/openshift/storage/OpenShiftServerTemplateStorageTest.java#L102

sutaakar commented on kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/main/java/org/kie/server/controller/openshift/storage/OpenShiftServerTemplateStorage.java:
```diff
@@ -0,0 +1,246 @@
+
+    @Override
+    public ServerTemplate delete(String identifier) {
+        // Can delete only detached CM, meaning not having an associated DC
```
Is this something to be implemented?

rhtevan replied:
The implementation is  complete. 
https://github.com/kiegroup/droolsjbpm-integration/blob/53dcf311f6ebcbb7ef58f03268ace7f987078ab4/kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/main/java/org/kie/server/controller/openshift/storage/OpenShiftServerTemplateStorage.java#L127-L131

sutaakar replied:
Ok, I was confused by the comment.
Maybe it would be better to place the comment into the javadoc, saying that UnsupportedOperationException is thrown in that case.

rhtevan replied:
Updated javadoc

sutaakar commented on kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/test/java/org/kie/server/controller/openshift/storage/OpenShiftServerTemplateStorageTest.java:
```diff
@@ -0,0 +1,208 @@
+    @Override
+    @Test
+    public void testStoreLoadAndUpdateServerTemplate() {
+        createDummyDC(serverTemplate.getId());
```
What is the DC used for here?

rhtevan replied:
The reason for adding a dummy DC is to simulate the attached CM scenario, meaning that any attempt to delete a CM/ServerTemplate which has already been used by KieServer Pod (A DC with same name exists) will be blocked by throwing an error.

A gap in the original test method implementation has been fixed by a new test method. 

https://github.com/rhtevan/droolsjbpm-integration/blob/fd4635e8e9b2263458db3339d7ea0c94b7ae87b6/kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/test/java/org/kie/server/controller/openshift/storage/OpenShiftServerTemplateStorageTest.java#L167-L177

sutaakar commented on kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/test/resources/test-kieserver-template.xml:
```diff
@@ -0,0 +1,44 @@
+<org.kie.server.controller.api.model.spec.ServerTemplate>
```
What is this file used for?

rhtevan replied:
Removed that deprecated test data file.

sutaakar commented on kie-server-parent/kie-server-controller/kie-server-controller-openshift/pom.xml:
```diff
@@ -0,0 +1,142 @@
+    <!-- fabric8 kubernetes and openshift java client dependencies -->
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>kubernetes-model</artifactId>
```
These Kubernetes/OpenShift artifacts are needed just for test purposes in this module. Can you give them a test scope (to keep dependency tree as clean as possible)?

rhtevan replied:
Updated those dependencies with test scope.

rhtevan commented:
> @rhtevan The OpenShiftServerTemplateStorageTest.testStoreExistingTemplateShouldThrowError is failing when running locally.

Removed that redundant test as the similar scenario has been covered at (https://github.com/rhtevan/droolsjbpm-integration/blob/041052b96b020cc64bd7bd84a9f09d56ec2defbc/kie-server-parent/kie-server-services/kie-server-services-openshift/src/test/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateOpenShiftRepositoryRegularTest.java#L292-L298)

sutaakar commented:
Jenkins retest this

rhtevan commented:
In response to newly identified java 11 compatibility issue (https://issues.jboss.org/browse/JBPM-8154), POM has been updated. (https://github.com/rhtevan/droolsjbpm-integration/blob/271eb36ed9f299d037ca980f63e77ff64f68830a/kie-server-parent/kie-server-controller/kie-server-controller-openshift/pom.xml#L101-L106)

rhtevan commented:
A key development of integrating Workbench and KieServer on OpenShiftStartupStrategy is to adapting ServerTemplate, which is BC/WB side data structure representing KieServer state, into KieServerState. To mitigate risk of compatibility issue, both ServerTemplate and KieServerState are being kept as-is. (No changes). On the flip side, it adds some challenges to accommodate all data items. For example, the following config-item(s) from KieServerState will/should not be converted into ServerTemplate and vice versa.  

          <config-item>
            <name>org.kie.server.persistence.ds</name>
            <value>java:/jboss/datasources/ExampleDS</value>
            <type>java.lang.String</type>
          </config-item>
          <config-item>
            <name>org.kie.server.repo</name>
            <value>/home/jboss/.kie/repository</value>
            <type>java.lang.String</type>
          </config-item>
          <config-item>
            <name>org.kie.server.sync.deploy</name>
            <value>true</value>
            <type>java.lang.String</type>
          </config-item>

As a result, it brought a subtle issue once KieServer is managed by BC/WB (Not apply to independent KieServer) in that those un-mapped system config items will get 'lost' from KieServerState ConfigMap. A simple solution has been implemented as follows:

https://github.com/rhtevan/droolsjbpm-integration/blob/7bee620d687f50f7103cc6422e559441a0a0e4f0/kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/KieServerConfig.java#L83

https://github.com/rhtevan/droolsjbpm-integration/blob/7bee620d687f50f7103cc6422e559441a0a0e4f0/kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/KieServerConfig.java#L87-L90

The value of those system level configuration items for KieServer should be retrieved from java system properties anyway.

rhtevan commented:
Minor update to log ServerTemplate cache TTL settings.

https://github.com/rhtevan/droolsjbpm-integration/blob/64f0a7a97235c7b0e02fb38b20fa17a026943b85/kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/main/java/org/kie/server/controller/openshift/storage/OpenShiftServerTemplateStorage.java#L51

rhtevan commented:
@sutaakar 
Updated KieServerStateFileInit and related classes so as to support immutable kieserver scenario and provide backward compatibility. 

https://github.com/rhtevan/droolsjbpm-integration/blob/51a045334009f4fc54321e158b3fe5fe005bbb2d/kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/storage/file/KieServerStateFileInit.java#L58-L78

sutaakar commented on kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/storage/KieServerStateRepositoryUtils.java:
```diff
@@ -36,10 +40,51 @@ public static void populateWithProperties(KieServerConfig config, Properties pro
             if (property.startsWith("org.kie.server") || property.startsWith("org.kie.executor")) {
-                KieServerConfigItem configItem = new KieServerConfigItem(property, properties.getProperty(property), String.class.getName());
+                KieServerConfigItem configItem = new KieServerConfigItem(
+                    property, properties.getProperty(property).replaceAll("^\"|\"$", ""), String.class.getName());
```
What is the purpose of this replacement?

sutaakar commented on kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateOpenShiftRepository.java:
```diff
@@ -106,10 +114,16 @@ public synchronized void store(@NotNull String serverId, @NotNull KieServerState
-                throw new IllegalStateException("KieServerState ConfigMap must exist before update.");
-            } else {
+            
+            if (Boolean.parseBoolean(kieServerState.getConfiguration()
```
Does this mean that for immutable Kie server a new CM is created/overridden but no rollout is triggered, for other cases data is updated and rollout is triggered, and for the last case - CM doesn't contain USED or IMMUTABLE labels or container deploying or undeploying is done in IMMUTABLE CM - nothing is done?
Do I understand this correctly?

rhtevan replied:
Once the immutable kie server CM has been created by the S2I/KieServerFileInit process, any update to it is not allowed and blocked.

rhtevan commented on kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateOpenShiftRepository.java:
```diff
@@ -185,6 +199,48 @@ public boolean isKieServerReady() {
+                    currentState.getContainers().size() == newState.getContainers().size()) {
+                /**
+                 * Deploy new or undeploy container(s) are not allowed for immutable kie server,
+                 * but changing container and scanner status is permitted.
```
Same above, immutable kie server state modification is not allowed.

rhtevan commented on kie-server-parent/kie-server-services/kie-server-services-openshift/src/test/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateOpenShiftRepositoryRegularTest.java:
```diff
@@ -329,6 +330,7 @@ public void testRetrieveAllKieServerIdsAndStates() {
     }
 
     @Test
+    @Ignore
```
Due to the limitation from f8's mock OpenShift server, this test case is disabled.

rhtevan commented:
@mswiderski Maciej, since you approved this PR few days ago, there were some further changes to support current immutable KieServer architecture. Appreciated if you could take a quick look and see if there were anything caught by your eye. @sutaakar Karel is actively verifying it again on OpenShift now. Once it's done, we will proceed to merge.

mswiderski commented:
thanks for bring this to my attention@rhtevan, all looks ok to me so just waiting for @sutaakar approval and we should be good to go.

rhtevan commented:
@sutaakar The code we had walked through together early today is in now, see if you have further comments. I'll address it ASAP.

rhtevan commented:
@sutaakar rebased with master

sutaakar commented on kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateOpenShiftRepository.java:
```diff
@@ -175,16 +186,101 @@ public KieServerState load(@NotNull String serverId) {
+        for (KieContainerResource container : state.getContainers()) {
+            if (container.getStatus().equals(KieContainerStatus.STOPPED) &&
+                newState.getContainers().stream()
+                        .filter(c -> c.getContainerId().equals(container.getContainerId()))
```
Could be replaced by .noneMatch(c -> c.getContainerId().equals(container.getContainerId())))

rhtevan replied:
Updated as per your suggestion

sutaakar commented on kie-server-parent/kie-server-services/kie-server-services-openshift/src/test/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateOpenShiftRepositoryRegularTest.java:
```diff
@@ -348,4 +352,64 @@ public void testRetrieveAllKieServerIdsAndStatesWithContaminatedCF() {
+        KieServerState stateWithStopped = (KieServerState) xs.fromXML(
+            cfmStopped.getData().get(KieServerStateCloudRepository.CFG_MAP_DATA_KEY));
+        
+        for (KieContainerResource container : stateWithStopped.getContainers()) {
```
Same as above

rhtevan replied:
Updated as per your suggestion

sutaakar commented:
Looks good apart from minor comments.
Need to check this also on real OpenShift instance. I hope I can manage to do it today, in the worst case in the Monday. If this becomes the only blocking thing from merge then don't wait for it (as I have already checked most part of PR in my previous checks).

rhtevan commented on kie-server-parent/kie-server-services/kie-server-services-openshift/src/test/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateOpenShiftRepositoryRegularTest.java:
```diff
@@ -348,4 +352,64 @@ public void testRetrieveAllKieServerIdsAndStatesWithContaminatedCF() {
+        KieServerState stateWithStopped = (KieServerState) xs.fromXML(
+            cfmStopped.getData().get(KieServerStateCloudRepository.CFG_MAP_DATA_KEY));
+        
+        for (KieContainerResource container : stateWithStopped.getContainers()) {
```
Updated as per your suggestion

sutaakar replied:
Thanks, btw what is this test used for? The only assert here checks condition which is set also in this test.

rhtevan replied:
Removed this redundant test

rhtevan commented:
@sutaakar Good suggestion, made the change accordingly. 
As this PR has not been impacted by the packaging issue, we can merge it now once you've done the final check. I'm updating my images shared with you on quay.io now.

rhtevan commented:
Just listed out dependent PR:
- https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/904

rhtevan commented:
@sutaakar 
Addressed optional issues caused by fidgeting at BC/WB during DC rollout. Added unit test and tested on local OpenShift clusters. Updating quay.io images shortly. 

https://github.com/rhtevan/droolsjbpm-integration/blob/8fd331cd338d44fbb461be3e60fb944a28b30958/kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateOpenShiftRepository.java#L214-L234

sutaakar commented on kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateOpenShiftRepository.java:
```diff
@@ -175,37 +186,128 @@ public KieServerState load(@NotNull String serverId) {
+        for (KieContainerResource container : state.getContainers()) {
+            if (container.getStatus().equals(KieContainerStatus.STARTED) &&
+                !newState.getContainers().stream()
+                         .noneMatch(c -> c.getContainerId().equals(container.getContainerId()) &&
```
Could you please use anyMatch() here and remove negation?
I think it would improve readability :)

rhtevan replied:
Good point, changed as per suggestion.

sutaakar commented:
Looks good to me.

sutaakar commented:
Jenkins retest this
